python/ml4ir/applications/ranking/data/scripts/create_dataset.py

```diff
@@ -0,0 +1,227 @@
+
+        # Build data
+        seed_data = load_seed_data(data_dir, logger)
+
+        df_synthetic = fill_data(logger,
+                                 seed_data,
+                                 max_num_records,
+                                 feature_config,
+                                 feature_highval,
+                                 feature_num_results,
+                                 num_samples)
+        file_io.write_df(df_synthetic, outfile=out_file, index=False)
+        logger.info('Synthetic data created! Location: {}'.format(out_file))
+        return df_synthetic
+
+    except Exception as e:
+        logger.error("!!! Error creating synthetic data: !!!\n{}".format(str(e)))
+        traceback.print_exc()
+        return
+
+
+def setup_logging():
+    return logging_utils.setup_logging(reset=True, file_name='test.txt', log_to_file=True)
+
+
+def set_seeds(random_state):
+    # for repeatability
+    np.random.seed(random_state)
+    random.seed(random_state)
+
+
+def load_seed_data(data_dir, logger):
+    # Start with loading in CSV format only
+    if file_io.path_exists(data_dir):
+        dfs = file_io.get_files_in_directory(data_dir)
+        return file_io.read_df_list(dfs)
+    else:
+        logger.error("Error! Data directory must exist and be specified")
+
+
+def generate_key(seed_dict, query_key, log=None):
+    seed_id = str(random.sample(seed_dict[query_key], 1)[0])
+    q_id = ''.join(random.sample(seed_id, len(seed_id)))
+    if q_id in set(seed_dict.keys()):
+        if log:
+            log.info('Generated query key is duplicate, regenerating. If happens repeatedly, query key may be too short')
+        return generate_key(seed_dict, query_key)
+    return q_id
+
+
+def filter_nres(seed_data, feature_query_key, feature_rank, feature_nres, max_num_records, logger):
+    nres = 'num_results_calc'
+    if nres in set(seed_data.columns):
+        nres = '{}_seed'.format(nres)
+    seed_data[nres] = seed_data.groupby(feature_query_key).transform('count')[feature_rank]
+
+    # If the number of results feature name was given, remove queries with not all results included
+    if feature_nres and feature_nres in set(seed_data.columns):
+        seed_data = seed_data.loc[(seed_data[feature_nres].astype('int64')==seed_data[nres])]
+    # Filter to max number of records if given
+    if max_num_records:
+        seed_data = seed_data.loc[seed_data[nres] <= max_num_records]
+
+    if len(seed_data) == 0:
+        logger.error('All seed data has been filtered out!')
+    return seed_data, nres
+
+
+def add_feature_highval(row_dict, feature_highval, is_labeled_pos, q_labelrank):
+    # Create ranking-specific high value scenario
+    # of the format {high_value_feature_1:[label1_val, label2_val]}
+    # Not creating examples of non-catastrophic failure, ie where the clicked record is not a name match
+    for hvk in feature_highval.keys():
+        if is_labeled_pos and q_labelrank > 1:
```

Comment:
@calliebradley [Minor] Can we better name `is_labeled_pos`. I don't follow what this is. You can use ranking specific terms like `click_rank` if that's what you are trying to do. Don't necessarily have to use `pos`. That was more of an abstraction for other tasks as well.

Reply:
Ok, no problem. It used to be `click_rank` but I had requests to make it more generalized. Changing it back to `click_rank`.